core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java

```diff
@@ -114,7 +117,12 @@ public boolean prebuild(Build<?,?> build, BuildListener listener) {
      *      Use {@link #perform(AbstractBuild, Launcher, BuildListener)} instead.
      */
     @Deprecated
-    public boolean perform(Build<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+    public boolean perform(Build<?, ?> build, Launcher launcher, BuildListener listener)
+            throws InterruptedException, IOException {       
+        if (build instanceof AbstractBuild && Util.isOverridden(BuildStepCompatibilityLayer.class, this.getClass(),
+                "perform", AbstractBuild.class, Launcher.class, BuildListener.class)) {
+            return perform((AbstractBuild<?, ?>) build, launcher, listener);
+        }
         throw new UnsupportedOperationException();
```

Comment:
BTW for newer refactorings I have done (such as `AbstractBuild` → `Run`), I have used `AbstractMethodError` to indicate cases where we have an introduced method that we wish we could have marked `abstract`.